### ISSUE TITLE
Pytorch Version and Windows Compatibility

### ISSUE THREAD

**Issue by bjascob** [bjascob/amrlib#1]
From Moha

Many thank for your quick response, I have done what you wrote,  I have
installed req file. now I am fighting with this torch compatibility problem,
I faced with this:

now I am fighting with this torch compatibility problem:
```
version_ <= kMaxSupportedFileFormatVersion INTERNAL ASSERT FAILED at
..\caffe2\serialize\inline_container.cc:132, please report a bug to PyTorch.
Attempted to read a PyTorch file with version 3, but the maximum supported
version for reading is 2. Your PyTorch installation may be too old. (init at
..\caffe2\serialize\inline_container.cc:132)
(no backtrace available)
```
it seems it related to versio0n of the torch, if you have  any idea would be happy to know that

**Comment by bjascob** [comment]
I'm not familiar with this error but I'm using torch v1.6.0.
I think the key is... "Attempted to read a PyTorch file with version 3, but the maximum supported version for reading is 2"
What version of torch are you using?  
I'm guessing that at some point they changed the file serialization protocol and older versions of torch can't read the newer.  I'm not sure when that happened.  Can you upgrade to torch 1.6?

**Comment by bjascob** [comment]
I found this buried in the torch 1.6 release notes...
_Deprecated old torch.save serialization format (#39460, #39893, #40288, #40793)
We have switched torch.save to use a zip file-based format by default rather than the old Pickle-based format. torch.load has retained the ability to load the old format, but use of the new format is recommended._ 

Since I use created the models with torch v1.6 I guess that's the minimum version.  I'll update the requirements file if you confirm this is the issue.

**Comment by myeghaneh** [comment]
thank you very much for opening this thread, I have worked a bit on that error, now I have 
```
torch              1.6.0+cu101
torchvision        0.6.0+cu92
```
then I faced with this: 
```
Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.
``` 
which is again related to the version of the torch 

I win 10 use , I used a clean python env based on your instruction, after installing req file and providing a folder 
data/model/ 
putting those model I still face with the error above, I have a feeling that it is related to the version of torch

**Comment by bjascob** [comment]
Do you have a GPU?  It looks like your pytorch version is setup for cuda 10.2 so I'm assuming you do.
The above error means torch doesn't think you have a GPU and it's trying to load the model to the GPU but can't.
Assuming you have a GPU try this from the python3 command prompt..
```
import torch
torch.cuda.is_available()
```
This should return True.  If not, there's something wrong with the torch/cuda setup.

**Comment by myeghaneh** [comment]
thank you again for the super quick response, 
here is the answer of that: 
```
import torch
torch.cuda.is_available()
False
``` 
basically you are saying that I should install another version of PyTorch? 
if yes. which one ?

**Comment by myeghaneh** [comment]
if Cuda 11, I see  that is not available for windows 

https://varhowto.com/install-pytorch-cuda-11-0/
:(

**Comment by bjascob** [comment]
Everything should work with cuda 10.1 which is what the pytorch version sees to be setup for.
I'm not sure what's wrong with your installation but the above test shows that it can't detect Cuda  and your GPU.  Something must not be installed properly.  Did you install Cuda.  What version?  Which GPU do you have?

**Comment by myeghaneh** [comment]
tnx for the response , should not be installed by  runing that req file? 
here is my pip list
```
amrlib             0.1.0
argon2-cffi        20.1.0
attrs              20.1.0
backcall           0.2.0
bleach             3.1.5
blis               0.4.1
catalogue          1.0.0
certifi            2020.6.20
cffi               1.14.2
chardet            3.0.4
click              7.1.2
colorama           0.4.3
cymem              2.0.3
decorator          4.4.2
defusedxml         0.6.0
entrypoints        0.3
filelock           3.0.12
future             0.18.2
graphviz           0.14.1
idna               2.10
ipykernel          5.3.4
ipython            7.18.1
ipython-genutils   0.2.0
ipywidgets         7.5.1
jedi               0.17.2
Jinja2             2.11.2
joblib             0.16.0
jsonschema         3.2.0
jupyter            1.0.0
jupyter-client     6.1.7
jupyter-console    6.2.0
jupyter-core       4.6.3
MarkupSafe         1.1.1
mistune            0.8.4
murmurhash         1.0.2
nbconvert          5.6.1
nbformat           5.0.7
nltk               3.5
notebook           6.1.3
numpy              1.19.1
packaging          20.4
panda              0.3.1
pandas             1.1.1
pandocfilters      1.4.2
parso              0.7.1
Penman             1.1.0
pickleshare        0.7.5
Pillow             7.2.0
pip                20.2.2
plac               1.1.3
preshed            3.0.2
prometheus-client  0.8.0
prompt-toolkit     3.0.7
pycparser          2.20
Pygments           2.6.1
pyparsing          2.4.7
PyQt5              5.15.0
PyQt5-sip          12.8.1
pyrsistent         0.16.0
python-dateutil    2.8.1
pytz               2020.1
pywin32            228
pywinpty           0.5.7
pyzmq              19.0.2
qtconsole          4.7.7
QtPy               1.9.0
regex              2020.7.14
requests           2.24.0
sacremoses         0.0.43
Send2Trash         1.5.0
sentencepiece      0.1.91
setuptools         49.6.0.post20200814
six                1.15.0
smatch             1.0.4
spacy              2.3.2
srsly              1.0.2
terminado          0.8.3
testpath           0.4.4
thinc              7.4.1
tokenizers         0.8.1rc2
torch              1.6.0+cu101
torchvision        0.6.0+cu92
tornado            6.0.4
tqdm               4.48.2
traitlets          5.0.2
transformers       3.1.0
Unidecode          1.1.1
urllib3            1.25.10
wasabi             0.8.0
wcwidth            0.2.5
webencodings       0.5.1
wheel              0.35.1
widgetsnbextension 3.5.1
wincertstore       0.
```

I have  NVIDIA GefForce GTX 980 Ti and AMD Radeopn R7 450

**Comment by bjascob** [comment]
Cuda is not installed with pip.  It's an NVida GPU only library that would need to be installed separately, however I'm not at all familiar with Conda so maybe that installed it for you.  In any event, the fact that `torch.cuda.is_available()` return `False` means it's not working.

Even though your cuda setup doesn't see to be working, with a small change the parse model loader you should be able to load to the CPU.  The change you'll need to make is in `amrlib/models/parse_gsii/inference.py`.  

Change the line `model_dict = torch.load(model_fpath)` to `model_dict = torch.load(model_fpath, map_location='cpu')`.  That should force the parser model to bypass the GPU and load to the cpu instead.  This will work for the parser model but the generation model is a little different and I'm not sure if it will have an issue loading or not.  If you still get an error, see what model it's trying to load when the error occurs.

**Comment by myeghaneh** [comment]
many thanks, :) I have changed inference.py ,now it works and it read the model, I saw those green processing bar,  now I i faced with this 
```
[E050] Can't find model 'en'. It doesn't seem to be a shortcut link, a Python package or a valid path to a data directory.
``` 
which is kind of familiar error for me, I guess I only need to load the model with "en_core_web_sm" insted of "en" 
am i right? 
I can not find in your script

**Comment by bjascob** [comment]
Sorry, I forgot to put in the instructions about manually downloading the spacy model.  I'll update the docs and be sure to use the long name.
The default value for this is in `amrlib/defaults.py`.  Just change that to `en_core_web_sm` or whatever spacy model you have.

**Comment by myeghaneh** [comment]
thank you very much again, It solved the problem  
sorry for bothering agian, i faced now with this 
```
Expected tensor for argument #1 'indices' to have scalar type Long; but got torch.IntTensor instead (while checking arguments for embedding)
```

**Comment by bjascob** [comment]
Can you print more the the error message and detail what exactly you're doing when this happened?  (are you running the GUI or another set of commands?).

**Comment by myeghaneh** [comment]
tnx, sure,  I am runing the sample in your repo 

```
import amrlib
stog = amrlib.load_stog_model()
graphs = stog.parse_sents(['This is a test of the system.', 'This is a second sentence.'])
for graph in graphs:
    print(graph)
``` 
then i faced with this
```
Expected tensor for argument #1 'indices' to have scalar type Long; but got torch.IntTensor instead (while checking arguments for embedding)
```

I am runing this on my jupyter notebook

**Comment by bjascob** [comment]
The examples run on Ubuntu from the command line.  I'm not sure what's causing this but it could be related to the map_location='cpu' change.  I'll try to replicate it, but I won't be able to dig into to it for a day or so.  I'll respond back when I have a solution or additional questions.

Is there any traceback with the error?  It would be helpful to know what line in the code this is causing this and what where the calls leading up to it.

**Comment by myeghaneh** [comment]
that is very nice of you , sure

```
Loading model C:\Users\moha\Anaconda3\envs\argenv\Lib\site-packages\amrlib\data\model_stog\model.pt
Loading SpaCy NLP Model: en_core_web_sm
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-565b383e0763> in <module>
      1 import amrlib
      2 stog = amrlib.load_stog_model()
----> 3 graphs = stog.parse_sents(['This is a test of the system.', 'This is a second sentence.'])
      4 for graph in graphs:
      5     print(graph)

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\amrlib\models\parse_gsii\inference.py in parse_sents(self, sents, add_metadata)
     63                 sio_f.write('\n')
     64         sio_f.seek(0)
---> 65         return self.parse_file_handle(sio_f, add_metadata)
     66 
     67     # parse a list of spacy spans (ie.. span has list of tokens)

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\amrlib\models\parse_gsii\inference.py in parse_file_handle(self, sio_f, add_metadata)
     91         for batch in data_loader:
     92             batch = move_to_device(batch, self.model.device)
---> 93             res = self._parse_batch(batch)
     94             for concept, relation in zip(res['concept'], res['relation']):
     95                 graph_lines = self.post_proc.postprocess(concept, relation)

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\amrlib\models\parse_gsii\inference.py in _parse_batch(self, batch)
    153         concept_batch = []
    154         relation_batch = []
--> 155         beams = self.model.work(batch, self.beam_size, self.max_time_step)
    156         score_batch = []
    157         for beam in beams:

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\amrlib\models\parse_gsii\modules\parser.py in work(self, data, beam_size, max_time_step, min_time_step)
     78         with torch.no_grad():
     79             if self.bert_encoder is not None:
---> 80                 word_repr, word_mask, probe = self.encode_step_with_bert(data['tok'], data['lem'],
     81                             data['pos'], data['ner'], data['word_char'], data['bert_token'], data['token_subword_index'])
     82             else:

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\amrlib\models\parse_gsii\modules\parser.py in encode_step_with_bert(self, tok, lem, pos, ner, word_char, bert_token, token_subword_index)
     60 
     61     def encode_step_with_bert(self, tok, lem, pos, ner, word_char, bert_token, token_subword_index):
---> 62         bert_embed, _ = self.bert_encoder(bert_token, token_subword_index=token_subword_index)
     63         word_repr = self.word_encoder(word_char, tok, lem, pos, ner)
     64         bert_embed = bert_embed.transpose(0, 1)

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\torch\nn\modules\module.py in _call_impl(self, *input, **kwargs)
    720             result = self._slow_forward(*input, **kwargs)
    721         else:
--> 722             result = self.forward(*input, **kwargs)
    723         for hook in itertools.chain(
    724                 _global_forward_hooks.values(),

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\amrlib\models\parse_gsii\bert_utils.py in forward(self, input_ids, token_subword_index)
     53         attention_mask = input_ids.ne(0)
     54 
---> 55         encoded_layers, pooled_output = super(BertEncoder, self).forward(
     56             input_ids, attention_mask, token_type_ids)
     57         if token_subword_index is None:

~\AppData\Roaming\Python\Python38\site-packages\transformers\modeling_bert.py in forward(self, input_ids, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds, encoder_hidden_states, encoder_attention_mask, output_attentions, output_hidden_states, return_dict)
    822         head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)
    823 
--> 824         embedding_output = self.embeddings(
    825             input_ids=input_ids, position_ids=position_ids, token_type_ids=token_type_ids, inputs_embeds=inputs_embeds
    826         )

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\torch\nn\modules\module.py in _call_impl(self, *input, **kwargs)
    720             result = self._slow_forward(*input, **kwargs)
    721         else:
--> 722             result = self.forward(*input, **kwargs)
    723         for hook in itertools.chain(
    724                 _global_forward_hooks.values(),

~\AppData\Roaming\Python\Python38\site-packages\transformers\modeling_bert.py in forward(self, input_ids, token_type_ids, position_ids, inputs_embeds)
    205 
    206         if inputs_embeds is None:
--> 207             inputs_embeds = self.word_embeddings(input_ids)
    208         position_embeddings = self.position_embeddings(position_ids)
    209         token_type_embeddings = self.token_type_embeddings(token_type_ids)

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\torch\nn\modules\module.py in _call_impl(self, *input, **kwargs)
    720             result = self._slow_forward(*input, **kwargs)
    721         else:
--> 722             result = self.forward(*input, **kwargs)
    723         for hook in itertools.chain(
    724                 _global_forward_hooks.values(),

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\torch\nn\modules\sparse.py in forward(self, input)
    122 
    123     def forward(self, input: Tensor) -> Tensor:
--> 124         return F.embedding(
    125             input, self.weight, self.padding_idx, self.max_norm,
    126             self.norm_type, self.scale_grad_by_freq, self.sparse)

c:\users\moha\anaconda3\envs\argenv\lib\site-packages\torch\nn\functional.py in embedding(input, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq, sparse)
   1812         # remove once script supports set_grad_enabled
   1813         _no_grad_embedding_renorm_(weight, input, max_norm, norm_type)
-> 1814     return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
   1815 
   1816 

RuntimeError: Expected tensor for argument #1 'indices' to have scalar type Long; but got torch.IntTensor instead (while checking arguments for embedding)
```

here is the complete error

**Comment by bjascob** [comment]
Ugly.  This looks to be way down in the pre-trained BERT embedding which is code I didn't write.  Since it doesn't happen on Ubuntu, it may be a Windows specific implementation issue (long vs int data-type).  Give me a few days.  I may need to find a Windows box somewhere to replicate this.

**Comment by myeghaneh** [comment]
I also work to find the problem, I found this thread close:
https://github.com/dmlc/dgl/issues/672

node_id is apparently an int32 tensor instead of int64 on Windows.

The problem is that np.array([1,2,3]) on Linux returns an int64 array, while on Windows it returns int32 even on an x64 machine. This is because the default numpy int inherits C long, which is 64-bit on Linux and 32-bit on Windows.

Putting dtype=int64 in either place (the data loader, returning int64 array for uniq_v in generate_sampled_graph_and_labels, or converting node_id to long) would fix the problem.

P.S. I guess from now on the best practice would be always specifying numpy dtype whenever we call np.array() just in case this mess happens. This whole thing is just ridiculous.

**Comment by bjascob** [comment]
I was able to get this working under Windows.  I won't check the changes in for a while.  I need to verify they don't mess-up anything in Linux plus I'll run some additional tests.

In case you want to try this for yourself, here's the changes to make.  It's just adding dtype='int64' in 5 places.
In `amrlib\models\parse_gsii\data_loader.py` change the following lines (line numbers are approximate)
```
line 54    data = np.transpose(np.array(ys, dtype=np.int64))
line 70    data = np.transpose(np.array(ys, dtype=np.int64), (1, 0, 2))
line 77   x = np.array([ list(x.shape) for x in xs], dtype=np.int64)
line 80   data = np.zeros(shape, dtype=np.int64)
line 115  _rel = np.full((1+out_conc_len, bsz, out_conc_len), vocabs['rel'].token2idx(PAD), dtype=np.int64)

```

**Comment by myeghaneh** [comment]
many many thanks, I have changed the script as you wrote:
it is working like a beautiful Benz, 

can you let me know how can I draw the AMR graph?

**Comment by bjascob** [comment]
These code changes will be implemented in version 0.1.1
